regnety/dataset/imagenet.py

```diff
@@ -0,0 +1,268 @@
+        """
+        image = tf.cast(tf.io.decode_jpeg(example["image"]), tf.float32)
+        height = 512
+        width = 512
+        filename = example["filename"]
+        label = example["label"]
+        synset = example["synset"]
+        return {
+            "image": image,
+            "height": height,
+            "width": width,
+            "filename": filename,
+            "label": label,
+            "synset": synset,
+        }
+
+    def _read_tfrecs(self):
+        """Function for reading and loading TFRecords into a tf.data.Dataset.
+
+        Returns:
+            A tf.data.Dataset
+        """
+        ds = tf.data.TFRecordDataset(self.tfrecs_filepath)
+        ds = ds.map(
+            lambda example: tf.io.parse_example(example, ImageNet.TFRECS_FORMAT)
+        )
+        ds = ds.map(lambda example: self.decode_example(example))
+        return ds
+
+    def _scale_and_center_crop(self, image, h, w, scale_size, final_size):
+        """Resizes image to given scale size and returns a center crop. Aspect
+        ratio is NOT maintained.
+
+        Args:
+            image: tensor of the image
+            h: initial height of image
+            w: initial width of image
+            scale_size: Size of image to scale to
+            final_size: Size of final image
+
+        Returns:
+            Tensor of shape (final_size, final_size, 3)
+        """
+        if w < h and w != scale_size:
+            w, h = tf.cast(scale_size, tf.int64), tf.cast(
+                ((h / w) * scale_size), tf.int64
+            )
+            im = tf.image.resize(image, (w, h))
+        elif h <= w and h != scale_size:
+            w, h = tf.cast(((h / w) * scale_size), tf.int64), tf.cast(
+                scale_size, tf.int64
+            )
+            im = tf.image.resize(image, (w, h))
+        else:
+            im = tf.image.resize(image, (w, h))
+        x = tf.cast((w - final_size) / 2, tf.int64)
+        y = tf.cast((h - final_size) / 2, tf.int64)
+        return im[y : (y + final_size), x : (x + final_size), :]
+
+    def random_sized_crop(self, example, min_area=0.08, max_iter=10):
+        """
+        Randomly chooses an area and aspect ratio and resizes the image to those
+        values.
+
+        Args:
+            example: A dataset example.
+            min_area: Minimum area of image to be used
+            max_iter: Maximum iteration to try before implementing scale and
+                center crop. Randomly genereted width and height must be lower
+                than original width and height before max_iter.
+
+        Returns:
+            Example of same format as TFRECS_FORMAT
+        """
+
+        h, w = tf.cast(example["height"], tf.int64), tf.cast(
+            example["width"], tf.int64
+        )
+        area = h * w
+        return_example = False
+        image = example["image"]
+        num_iter = tf.constant(0, dtype=tf.int32)
+        while num_iter <= max_iter:
```

Comment:
Looping like this for randomization is a little wasteful and error-prone, especially since this will run many times. It's always best to avoid this if you can. Also note that our final image is always square, therefore we only need to choose one "size" to extract.

Consider a simplified approach, like:

```python
min_crop_size = math.sqrt(min_area * h * w)
# assuming you allow padding:
max_crop_size = max(h, w)
crop_size = tf.random.uniform((), minval=min_crop_size, maxval=max_crop_size, dtype=tf.int64)
# Note: you might need to pad or crop the image to make it square before calling random_crop.
crop = tf.image.random_crop(image, size=(crop_size, crop_size, 3))

# Now resize if necessary (assuming this function should always return images with a constant size)
```

Reply:
Yes, the approach seemed off to me, too, but I decided to stick with it as it was in the official repo. 

As far as I understand, the objective of this function is to distort the scale of the images (i.e. to stretch or squeeze the images) and then take a random crop. That's why I avoided smart_resize or any other method which preserves the scale. Please correct me if I am wrong.